src/datagram.cpp

```diff
@@ -3,73 +3,257 @@
+        did.len = 2;
+
+        auto& dat = bufs.emplace_back();
+        dat.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.second.first.data()));
+        dat.len = b.second.first.size();
+
+        return bufs;
     }
 
-    std::vector<ngtcp2_vec> DatagramIO::pending()
+    prepared_datagram DatagramIO::pending_datagram()
     {
         log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
 
-        std::vector<ngtcp2_vec> bufs{};
-        bufs.reserve(std::distance(user_buffers.begin(), user_buffers.end()));
+        std::pair<uint64_t, std::vector<ngtcp2_vec>> bufs;
+        bufs.second.reserve(2);
+
+        auto& b = send_buffer.front();
+
+        prepared_datagram d{};
+        d.id = b.first;
+        d.bufs_len = 1;
 
-        for (auto& b : user_buffers)
+        oxenc::write_host_as_big(static_cast<uint16_t>(b.first), d.dgid.data());
+
+        if (_packet_splitting)
         {
-            auto& temp = bufs.emplace_back();
+            d.bufs[0].base = d.dgid.data();
+            d.bufs[0].len = 2;
+            d.bufs_len++;
+        }
+
+        d.bufs[d.bufs_len - 1].base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.second.first.data()));
+        d.bufs[d.bufs_len - 1].len = b.second.first.size();
+
+        return d;
+    }
+
+    std::optional<bstring> DatagramIO::to_buffer(bstring_view data, uint16_t dgid)
+    {
+        log::trace(log_cat, "DatagramIO handed datagram with endian swapped ID: {}", dgid);
 
-            temp.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.first.data()));
-            temp.len = b.first.size();
+        auto r = recv_buffer.receive(data, dgid);
+
+        if (r)
+            return r;
+
+        return std::nullopt;
+    }
+
+    std::optional<bstring> rotating_buffer::receive(bstring_view data, uint16_t dgid)
+    {
+        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
+
+        {
+            std::lock_guard<std::mutex> buffer_lock(m);
+
+            auto idx = dgid >> 2;
+
+            row = (idx % bufsize) / rowsize;
+            col = idx % rowsize;
+
+            auto& b = buf.at(row).at(col);
+
+            if (not b.empty())
+            {
+#ifndef NDEBUG
+                if (enable_test_features)
+                {
+                    log::debug(log_cat, "enable_test_features is true, inducing packet loss");
+                    test_counter += 1;
+                    log::debug(log_cat, "test counter: {}", test_counter);
+                    return std::nullopt;
+                }
+                else
+                {
+                    log::debug(log_cat, "enable_test_features is false, skipping optional logic");
+                }
+#endif
+
+                log::trace(log_cat, "Pairing datagram (ID: {}) with counterpart at buffer pos [{},{}]", dgid, row, col);
+
+                bstring out{};
+                out.reserve(data.size() + b.data.size());
```

Comment:
I think that we could potentially avoid another string allocation here by reusing either `b.data` or `data`.  Maybe something like:

```C++
bstring out;
if (order) {
    b.data.append(data);
    out = std::move(b.data);
} else {
    data.append(b.data);
    out = std::move(data);
}
```

Then we could *also* reserve space when we first set up the partial `.data` element so that this append won't need to reallocate.

Reply:
No, wait, the second version here is wrong.

Reply:
v2:
```C++
if (order)
    b.data.append(data);
else {
    b.data.reserve(b.data.size() + data.size());
    // We have the second half and need to prepend the first half, so move it to the end:
    std::memmove(b.data.data() + data.size(), b.data.data(), b.data.size());
    // Now copy the first half into place:
    std::memcpy(b.data.data(), data.data(), data.size());
}
bstring out{std::move(b.data)};
```
which should need no new allocation *if* we pre-reserved the needed amount when creating the data.

Reply:
Also those two lines have a lot of Data.  Lore will be upset.

Reply:
quick note about the memmoves: when we already have the first half of the datagram, the incoming smaller half is overwriting the big half. this is with the new storage structures, so ill adjust and let you know

Reply:
The `reserve` should be a `resize` instead.